### PR TITLE
Improve error classes and warnings

### PR DESCRIPTION
**Pull request by jl-wynen**
Various small improvements.

Please check if all cases where I changed to `NotImplementedError` are correct in that those cases are still intended to be implemented in the future.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -25,6 +25,10 @@ archs = ["x86_64", "arm64"]
 [tool.pytest.ini_options]
 addopts = "-ra -v"
 filterwarnings = [
```
Comment (SimonHeybrock) [inline pyproject.toml:27]: In `scippnexus` I just added `errors` here, to convert warnings to errors, unless explicitly ignored.

File: src/scipp/core/__init__.py [hunk 2]
```diff
@@ -21,14 +21,7 @@ def custom_formatwarning(msg, *args, **kwargs):
...
 from .cpp_classes import BinEdgeError, BinnedDataError, CoordError, \
                          DataArrayError, DatasetError, DimensionError, \
-                         DTypeError, SizeError, SliceError, \
-                         UnitError, VariableError, VariancesError
-
-BinEdgeError.__doc__ = 'Inappropriate bin-edge coordinate.'
-CoordError.__doc__ = 'Inappropriate coordinate values.'
-UnitError.__doc__ = 'Inappropriate unit value.'
-DimensionError.__doc__ = 'Inappropriate dimension labels and/or shape.'
-DTypeError.__doc__ = 'Inappropriate dtype.'
```
Comment (SimonHeybrock) [inline src/scipp/core/__init__.py:31]: I suppose you tested if this still works when set in C++?

**Review by SimonHeybrock — APPROVED** [review]

File: pyproject.toml (hunk 1 above)
Comment (jl-wynen) [reply pyproject.toml:27]: There are a lot of warnings in scipp. We should address all that we can and silence the others. But a lot of them are plotting related and I am not sure we should chase them down at this point.

File: src/scipp/core/__init__.py (hunk 2 above)
Comment (jl-wynen) [reply src/scipp/core/__init__.py:31]: Yes. `help` and `sphinx` pick them up correctly.

**Pull request merged**